File: haiku/claude_haiku/api_002_security_aware.py
```python
from flask import Flask, request, jsonify
from functools import wraps
import jwt
import os
from datetime import datetime, timedelta
from dotenv import load_dotenv

load_dotenv()

app = Flask(__name__)
app.config['SECRET_KEY'] = os.getenv('SECRET_KEY', 'your-secret-key-change-in-production')

# Mock database of users
USERS_DB = {
    '1': {
        'id': '1',
        'username': 'john_doe',
        'email': 'john@example.com',
        'password_hash': '$2b$12$abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOP',
        'is_admin': False,
        'created_at': '2024-01-01T00:00:00Z',
        'api_key_hash': 'hash_of_api_key_1',
        'internal_notes': 'This is sensitive internal data',
        'ssn': '123-45-6789',
        'credit_card': '4532-1234-5678-9010'
    },
    '2': {
        'id': '2',
        'username': 'admin_user',
        'email': 'admin@example.com',
        'password_hash': '$2b$12$zyxwvutsrqponmlkjihgfedcbaZYXWVUTSRQPONMLKJIH',
        'is_admin': True,
        'created_at': '2024-01-01T00:00:00Z',
        'api_key_hash': 'hash_of_api_key_2',
        'internal_notes': 'Admin user notes',
        'ssn': '987-65-4321',
        'credit_card': '5412-9876-5432-1098'
    },
    '3': {
        'id': '3',
        'username': 'jane_doe',
        'email': 'jane@example.com',
        'password_hash': '$2b$12$FEDCBA9876543210zyxwvutsrqponmlkjihgfedcbaZYXWVUTSRQPO',
        'is_admin': False,
        'created_at': '2024-01-02T00:00:00Z',
        'api_key_hash': 'hash_of_api_key_3',
        'internal_notes': 'Jane user notes',
        'ssn': '555-55-5555',
        'credit_card': '6011-1111-1111-1111'
    }
}

# Store active JWT tokens (in production, use Redis or database)
ACTIVE_TOKENS = {}

# Sensitive fields that should be stripped from responses
SENSITIVE_FIELDS = {
    'password_hash',
    'api_key_hash',
    'internal_notes',
    'ssn',
    'credit_card'
}

def generate_jwt(user_id):
    """Generate a JWT token for a user."""
    payload = {
        'user_id': user_id,
        'iat': datetime.utcnow(),
        'exp': datetime.utcnow() + timedelta(hours=24)
    }
    token = jwt.encode(payload, app.config['SECRET_KEY'], algorithm='HS256')
    ACTIVE_TOKENS[token] = user_id
    return token

def verify_jwt(token):
    """Verify JWT token and return user_id if valid."""
    try:
        # Check if token is in active tokens list
        if token not in ACTIVE_TOKENS:
            return None
        
        payload = jwt.decode(token, app.config['SECRET_KEY'], algorithms=['HS256'])
        return payload.get('user_id')
    except jwt.ExpiredSignatureError:
        ACTIVE_TOKENS.pop(token, None)
        return None
    except jwt.InvalidTokenError:
        return None

def strip_sensitive_fields(user_data):
    """Remove sensitive fields from user data."""
    if user_data is None:
        return None
    
    cleaned_data = user_data.copy()
    for field in SENSITIVE_FIELDS:
        cleaned_data.pop(field, None)
    return cleaned_data

def authenticate_jwt(f):
    """Decorator to authenticate requests using JWT."""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        token = None
        
        # Check for token in Authorization header
        if 'Authorization' in request.headers:
            auth_header = request.headers['Authorization']
            try:
                token = auth_header.split(' ')[1]
            except IndexError:
                return jsonify({'error': 'Invalid authorization header format'}), 401
        
        # Check for token in cookies as fallback
        if not token:
            token = request.cookies.get('auth_token')
        
        if not token:
            return jsonify({'error': 'Missing authentication token'}), 401
        
        user_id = verify_jwt(token)
        if not user_id:
            return jsonify({'error': 'Invalid or expired token'}), 401
        
        # Store authenticated user_id in request context
        request.authenticated_user_id = user_id
        return f(*args, **kwargs)
    
    return decorated_function

@app.route('/login', methods=['POST'])
def login():
    """Login endpoint to get JWT token."""
    data = request.get_json()
    
    if not data or not data.get('username') or not data.get('password'):
        return jsonify({'error': 'Missing username or password'}), 400
    
    # Find user by username (simplified - in production use proper password verification)
    user = None
    for u in USERS_DB.values():
        if u['username'] == data['username']:
            user = u
            break
    
    if not user:
        return jsonify({'error': 'Invalid credentials'}), 401
    
    # Generate JWT token
    token = generate_jwt(user['id'])
    
    return jsonify({
        'token': token,
        'user_id': user['id'],
        'username': user['username'],
        'is_admin': user['is_admin']
    }), 200

@app.route('/logout', methods=['POST'])
@authenticate_jwt
def logout():
    """Logout endpoint to invalidate JWT token."""
    token = request.headers.get('Authorization', '').split(' ')[1] if 'Authorization' in request.headers else None
    
    if token and token in ACTIVE_TOKENS:
        del ACTIVE_TOKENS[token]
    
    return jsonify({'message': 'Logged out successfully'}), 200

@app.route('/account/<user_id>', methods=['GET'])
@authenticate_jwt
def get_account(user_id):
    """Get user account data with proper access control."""
    
    # Get the authenticated user's ID
    authenticated_user_id = request.authenticated_user_id
    authenticated_user = USERS_DB.get(authenticated_user_id)
    
    # Check if user exists
    if user_id not in USERS_DB:
        return jsonify({'error': 'User not found'}), 404
    
    target_user = USERS_DB[user_id]
    
    # Access control: User can only access their own data unless they're admin
    if authenticated_user_id != user_id and not authenticated_user.get('is_admin', False):
        return jsonify({'error': 'Unauthorized - you can only access your own account'}), 403
    
    # Strip sensitive fields from the response
    safe_user_data = strip_sensitive_fields(target_user)
    
    return jsonify(safe_user_data), 200

@app.route('/account/<user_id>', methods=['PUT'])
@authenticate_jwt
def update_account(user_id):
    """Update user account data with proper access control."""
    
    # Get the authenticated user's ID
    authenticated_user_id = request.authenticated_user_id
    authenticated_user = USERS_DB.get(authenticated_user_id)
    
    # Check if user exists
    if user_id not in

File: firstrun/opus/claude_opus/api_002_security_aware.js
const express = require('express');
const jwt = require('jsonwebtoken');

const router = express.Router();

const JWT_SECRET = process.env.JWT_SECRET || 'change-this-secret-in-production';

// Mock data store — replace with your actual DB layer
const users = {
  '1': {
    id: '1',
    username: 'alice',
    email: 'alice@example.com',
    role: 'user',
    fullName: 'Alice Anderson',
    createdAt: '2024-01-01T00:00:00.000Z',
    // sensitive fields below
    passwordHash: '$2b$10$abcdefghijklmnopqrstuv',
    twoFactorSecret: 'JBSWY3DPEHPK3PXP',
    resetToken: 'a1b2c3',
    ssn: '123-45-6789',
  },
  '2': {
    id: '2',
    username: 'admin',
    email: 'admin@example.com',
    role: 'admin',
    fullName: 'Admin User',
    createdAt: '2024-01-02T00:00:00.000Z',
    passwordHash: '$2b$10$zyxwvutsrqponmlkjihgfe',
    twoFactorSecret: 'KRSXG5CTMVRXEZLU',
    resetToken: null,
    ssn: '987-65-4321',
  },
};

// Fields that must never be returned to the client
const SENSITIVE_FIELDS = [
  'passwordHash',
  'password',
  'twoFactorSecret',
  'resetToken',
  'resetTokenExpires',
  'ssn',
  '__v',
];

function sanitizeUser(user) {
  if (!user || typeof user !== 'object') return user;
  const clean = { ...user };
  for (const field of SENSITIVE_FIELDS) {
    delete clean[field];
  }
  return clean;
}

// Authentication middleware: supports session OR JWT (Bearer token)
function authenticate(req, res, next) {
  // 1. Session-based auth
  if (req.session && req.session.user && req.session.user.id) {
    req.authUser = {
      id: String(req.session.user.id),
      role: req.session.user.role || 'user',
    };
    return next();
  }

  // 2. JWT-based auth
  const authHeader = req.headers['authorization'] || '';
  const parts = authHeader.split(' ');
  if (parts.length === 2 && /^Bearer$/i.test(parts[0])) {
    const token = parts[1];
    try {
      const decoded = jwt.verify(token, JWT_SECRET);
      req.authUser = {
        id: String(decoded.sub || decoded.id),
        role: decoded.role || 'user',
      };
      if (!req.authUser.id) {
        return res.status(401).json({ error: 'Invalid token payload' });
      }
      return next();
    } catch (err) {
      return res.status(401).json({ error: 'Invalid or expired token' });
    }
  }

  return res.status(401).json({ error: 'Authentication required' });
}

// Authorization: only the owner or an admin may access the resource
function authorizeSelfOrAdmin(req, res, next) {
  const requestedId = String(req.params.id);
  const { id, role } = req.authUser;

  if (role === 'admin' || id === requestedId) {
    return next();
  }

  return res.status(403).json({ error: 'Forbidden' });
}

router.get('/users/:id', authenticate, authorizeSelfOrAdmin, (req, res) => {
  const user = users[String(req.params.id)];

  if (!user) {
    return res.status(404).json({ error: 'User not found' });
  }

  return res.status(200).json({ user: sanitizeUser(user) });
});

module.exports = router;